### ISSUE TITLE
Configure errors with autoconf 2.70 

### ISSUE THREAD

**Issue by cniethammer** [open-mpi/ompi#8377]
Using autoconf 2.70 the configure script generated by autogen.pl is broken and aborts with
```./configure
./configure: line 3497: 5: Bad file descriptor
checking whether  is declared... ./configure: line 3499: ${+y}: bad substitution
```
Inspecting the generated configure script there seem to be a broken function definition for `ac_fn_c_check_decl` around the reported lines messing up brackets/configure code. A second, non-broken definition of `ac_fn_c_check_decl` occures later in the configure script... looking very strange to me.

I could not identify the exact cause for this as the generated configure is close to 0.5 million lines of code...

It could be
- either an issue in autoconf 2.70 as I see commits mentioning AC_CHECK_DECL (see. https://github.com/autotools-mirror/autoconf/commit/82ef7805faffa151e724aa76c245ec590d174580 and https://github.com/autotools-mirror/autoconf/commit/ec90049dfcf4538750e61d675d885157fa5ca7f8)
- or an Open MPI build system issue with future autoconf versions, as 
  there are a lot of warnings about obsolete macro uses as well as overrides for Automake targets by autogen.pl/autoconf 2.70.
```
configure.ac:54: warning: The macro `AC_CONFIG_HEADER' is obsolete.
configure.ac:466: warning: The macro `AC_HAVE_FUNCS' is obsolete.
configure.ac:593: warning: The macro `AC_FD_CC' is obsolete.
configure.ac:687: warning: The macro `AC_TRY_LINK' is obsolete.
configure.ac:687: warning: The macro `AC_TRY_COMPILE' is obsolete.
configure.ac:728: warning: The macro `AC_TRY_COMPILE' is obsolete.
...
ompi/mpi/c/profile/Makefile.am:462: warning: user target 'ID' defined here ...
/usr/share/automake-1.16/am/tags.am: ... overrides Automake target 'ID' defined here
ompi/mpi/c/profile/Makefile.am:459: warning: user target 'tags' defined here ...
/usr/share/automake-1.16/am/tags.am: ... overrides Automake target 'tags' defined here
...
```
###  Tested MPI Versions and environment
v4.1.0, master (228fc4147f08a7048845ce2f563587a36162e366)

* autoconf 2.70 + m4-1.4.18
* 34fa2ca7bea3e6d4e2a40b5ca37a5b3401d41615 3rd-party/openpmix (v4.0.0-7-g34fa2ca7)
* 4d7ba56f9e655b5d51b720c52b5839dcc3dac207 3rd-party/prrte (dev-30813-g4d7ba56f9e)
* CentOS 7.2 and arch linux on x86_64

### Steps to reproduce

```
tar xf openmpi-4.1.0.tar.bz2
cd openmpi-4.1.0/
autoconf --version # autoconf (GNU Autoconf) 2.70 ...
./autogen.pl --force
./configure
```

**Comment by jsquyres** [comment]
Looks like Autoconf 2.70 was released in Dec of 2020: https://lists.gnu.org/archive/html/autotools-announce/2020-12/msg00001.html, which contains this:

> Unfortunately, we were not able to maintain perfect backward compatibility with existing Autoconf scripts.  Caution is advised when upgrading.  The list of changes, below, includes detailed explanations and advice for all of the compatibility problems we know about.

I'm guessing you ran into some of the backwards incompatibilities.

This might be painful.  Ugh!

**Comment by cniethammer** [comment]
I just gave the latest developer version of autoconf a try (https://github.com/autotools-mirror/autoconf/commit/66fdaea3cfb4e758212c1891913e9a59441d49af) as one of the commits around AC_CHECK_DECL was after the release of autoconf 2.70 (https://github.com/autotools-mirror/autoconf/commit/ec90049dfcf4538750e61d675d885157fa5ca7f8)

This runs for some lines but breaks as well:
```
...
checking if --with-cuda is set... not set (--with-cuda=)
./configure: line 10688: syntax error near unexpected token `)'
./configure: line 10688: `    )
```

**Comment by wreckdump** [comment]
I am having the same issue. Any plans on updating the autogen.pl with up-to-date autoconf macros..? I would guess this would be a pain...

**Comment by jsquyres** [comment]
@wreckdump It's being worked on, but slowly (i.e., not a high priority).  Given that Autoconf deliberately introduced backwards incompatibilities, I think downstream distro packagers are going to have a heckuva time figuring out when to upgrade their bundled copies of Autoconf.  ☹️ 

For the moment, it would be best to simply use an older Autoconf.

**Comment by cniethammer** [comment]
The fix for this can be found in PR #8393.
A bit of cleanup of some obsolete macro warnings along the way can be found in #8392

**Comment by wreckdump** [comment]
@cniethammer It works! I'll also look into the cleanup. Much obliged!

**Comment by cniethammer** [comment]
Fix #8393 was merged into master
